### PR TITLE
docs: translate `Components and Hooks must be pure`

### LINKED ISSUE #652: Rules of React - Components and Hooks must be pure
Page: [id.react.dev/reference/rules/components-and-hooks-must-be-pure](https://id.react.dev/reference/rules/components-and-hooks-must-be-pure)

Assignee: @FelixFern

### PR DESCRIPTION
**Pull request by FelixFern**
Closes #652

## Description

Translate the reference/rules/components-and-hooks-must-be-pure
Page URL: https://id.react.dev/reference/rules/components-and-hooks-must-be-pure

### CODE REVIEW

**Comment by FelixFern** [comment]
ready for review @r17x @resir014

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 1]
```diff
@@ -1,44 +1,43 @@
...
+pe-*renderan* mengacu pada perhitungan seperti apa tampilan UI anda nantinya. Setelah me-*render*, [Efek](/reference/react/useEffect) di *flush* (artinya mereka akan dijalankan hingga tidak ada lagi yang tersisa) dan dapat memperbarui kalkulasi jika Efek berdampak pada *layout*. React akan mengambil kalkulasi ini dan membandingkannya dengan kalkulasi yang digunakannya pada versi sebelumnya dari UI anda, lalu *commits* hanya perubahan minim yang diperlukan ke [DOM](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model) (apa yang sebenarnya pengguna lihat) untuk menyesuaikan dengan versi terbaru.
```
Comment (resir014) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:30]:
```suggestion
pe-*renderan* mengacu pada perhitungan seperti apa tampilan UI anda nantinya. Setelah me-*render*, [*Effect*](/reference/react/useEffect) di *flush* (artinya mereka akan dijalankan hingga tidak ada lagi yang tersisa) dan dapat memperbarui kalkulasi jika *Effect* berdampak pada *layout*. React akan mengambil kalkulasi ini dan membandingkannya dengan kalkulasi yang digunakannya pada versi sebelumnya dari UI anda, lalu *commits* hanya perubahan minim yang diperlukan ke [DOM](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model) (apa yang sebenarnya pengguna lihat) untuk menyesuaikan dengan versi terbaru.
```

Keep *Effect* (as in React's Effects from `useEffect`) untranslated, as opposed to "efek samping" (the programming concept of side effects).

([Reference](https://id.react.dev/learn/synchronizing-with-effects#what-are-effects-and-how-are-they-different-from-events))

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 2]
```diff
@@ -59,7 +58,7 @@ function Dropdown() {
 function Dropdown() {
   const selectedItems = new Set();
   useEffect(() => {
-    // this code is inside of an Effect, so it only runs after rendering
+    // code ini berada didalam sebuah Efek, sehingga hanya akan jalan saat setelah proses render
```
Comment (resir014) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:61]:
```suggestion
    // kode ini berada didalam sebuah Effect, sehingga hanya akan jalan saat setelah proses render
```

Idem.

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 3]
```diff
@@ -116,133 +115,133 @@ export default function Clock() {
...
+[Efek samping](/learn/keeping-components-pure#side-effects-unintended-consequences) tidak seharusnya jalan [pada *render*](#how-does-react-run-your-code), karena React dapat *render* komponen-komponen beberapa kali untuk menghasilkan pengalaman pengguna sebaik mungkin.
 
 <Note>
-Side effects are a broader term than Effects. Effects specifically refer to code that's wrapped in `useEffect`, while a side effect is a general term for code that has any observable effect other than its primary result of returning a value to the caller.
+Efek samping adalah istilah yang lebih luas dari Efek. Efek secara khusus merujuk pada kode yang dibungkus dengan `useEffect`, sedangkan efek samping adalah istilah umum untuk kode yang memiliki efek yang dapat diamati selain dari hasil utamanya yaitu mengembalikan sebuah nilai kepada pemanggil.
 
-Side effects are typically written inside of [event handlers](/learn/responding-to-events) or Effects. But never during render.
+Efek samping biasanya ditulis di dalam [*event handler*](/learn/responing-to-events) atau Efek. Tetapi tidak pernah selama *render*.
```
Comment (resir014) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:131]:
```suggestion
Efek samping biasanya ditulis di dalam [*event handler*](/learn/responing-to-events) atau *Effect*. Tetapi tidak pernah selama *render*.
```

Idem.

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 4]
```diff
@@ -68,40 +67,40 @@ function Dropdown() {
...
+`new Date()` tidak idemponten karena selalu mengeluarkan tanggal saat ini dan hasilnya selalu berubah setiap kali dipanggil. Saat anda *render* komponen di atas, waktu yang ditampilkan pada layar akan tetap pada waktu dimana komponen tersebut di-*render*. Sama halnya dengan `Math.random()` yang juga tidak idempoten, karena selalu mengeluarkan nilai yang berbeda setiap kali dipanggil, walaupun masukan yang diberikan sama.
 
-This doesn't mean you shouldn't use non-idempotent functions like `new Date()` _at all_ – you should just avoid using them [during render](#how-does-react-run-your-code). In this case, we can _synchronize_ the latest date to this component using an [Effect](/reference/react/useEffect):
+Hal ini bukan berarti anda tidak seharusnya menggunakan fungsi idemponen seperti `new Date()` *sama sekali* - anda hanya perlu untuk menghindarinya [saat *render*](#how-does-react-run-your-code), Dalam kasus ini, kita bisa *menyinkronkan* tanggal terbaru ke komponen ini dengan [Efek](/reference/react/useEffect):
```
Comment (resir014) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:85]:
```suggestion
Hal ini bukan berarti anda tidak seharusnya menggunakan fungsi idemponen seperti `new Date()` *sama sekali* - anda hanya perlu untuk menghindarinya [saat *render*](#how-does-react-run-your-code), Dalam kasus ini, kita bisa *menyinkronkan* tanggal terbaru ke komponen ini dengan [*Effect*](/reference/react/useEffect):
```

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 5]
```diff
@@ -116,133 +115,133 @@ export default function Clock() {
 
 </Sandpack>
 
-By wrapping the non-idempotent `new Date()` call in an Effect, it moves that calculation [outside of rendering](#how-does-react-run-your-code).
+Dengan membungkus pemanggilan `new Date()` yang tidak idempoten di dalam sebuah Efek, ini akan memindahkan kalkulasi tersebut [di luar pe-*render*-an](#how-does-react-run-your-code).
```
Comment (resir014) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:118]:
```suggestion
Dengan membungkus pemanggilan `new Date()` yang tidak idempoten di dalam sebuah *Effect*, ini akan memindahkan kalkulasi tersebut [di luar pe-*render*-an](#how-does-react-run-your-code).
```

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 6]
```diff
@@ -116,133 +115,133 @@ export default function Clock() {
...
+## Efek samping seharusnya berjalan diluar *render* {/*side-effects-must-run-outside-of-render*/}
 
-[Side effects](/learn/keeping-components-pure#side-effects-unintended-consequences) should not run [in render](#how-does-react-run-your-code), as React can render components multiple times to create the best possible user experience.
+[Efek samping](/learn/keeping-components-pure#side-effects-unintended-consequences) tidak seharusnya jalan [pada *render*](#how-does-react-run-your-code), karena React dapat *render* komponen-komponen beberapa kali untuk menghasilkan pengalaman pengguna sebaik mungkin.
 
 <Note>
-Side effects are a broader term than Effects. Effects specifically refer to code that's wrapped in `useEffect`, while a side effect is a general term for code that has any observable effect other than its primary result of returning a value to the caller.
+Efek samping adalah istilah yang lebih luas dari Efek. Efek secara khusus merujuk pada kode yang dibungkus dengan `useEffect`, sedangkan efek samping adalah istilah umum untuk kode yang memiliki efek yang dapat diamati selain dari hasil utamanya yaitu mengembalikan sebuah nilai kepada pemanggil.
```
Comment (resir014) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:129]:
```suggestion
Efek samping adalah istilah yang lebih luas dari *Effect*. *Effect* secara khusus merujuk pada kode yang dibungkus dengan `useEffect`, sedangkan efek samping adalah istilah umum untuk kode yang memiliki efek yang dapat diamati selain dari hasil utamanya yaitu mengembalikan sebuah nilai kepada pemanggil.
```

**Review by resir014 — CHANGES_REQUESTED** [review]
One minor fix. @FelixFern

**Review by r17x — APPROVED** [review]

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 7]
```diff
@@ -307,35 +306,35 @@ function useIconStyle(icon) {
...
 ---
 
-## Values are immutable after being passed to JSX {/*values-are-immutable-after-being-passed-to-jsx*/}
+## Nilai tidak dapat diubah setelah diteruskan ke JSX {/*values-are-immutable-after-being-passed-to-jsx*/}
 
-Don't mutate values after they've been used in JSX. Move the mutation before the JSX is created.
+Jangan melakukan mutasi nilai setelah nilai tersebut digunakan dalam JSX. Pindahkan mutasi sebelum JSX dibuat.
 
-When you use JSX in an expression, React may eagerly evaluate the JSX before the component finishes rendering. This means that mutating values after they've been passed to JSX can lead to outdated UIs, as React won't know to update the component's output.
+Ketika Anda menggunakan JSX dalam sebuah ekspresi, React mungkin akan mengevaluasi JSX sebelum komponen selesai di-*render*. Ini berarti bahwa mengubah nilai setelah nilai tersebut dioper ke JSX dapat menyebabkan UI yang ketinggalan zaman, karena React tidak akan tahu untuk memperbarui keluaran komponen.
```
Comment (r17x) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:331]:
```suggestion
Ketika Anda menggunakan JSX dalam sebuah ekspresi, React mungkin akan mengevaluasi JSX sebelum komponen selesai di-*render*. Ini berarti bahwa mengubah nilai setelah nilai tersebut dioper ke JSX dapat menyebabkan UI yang sudah usang, Karena React tidak akan tahu untuk memperbarui keluaran komponen.
```

**Review by r17x — CHANGES_REQUESTED** [review]
Minor

Dengan menggunakan kata "sudah usang" lebih dekat dengan konteks dari "outdates UIs".

**Review by resir014 — APPROVED** [review]

File: src/content/reference/rules/components-and-hooks-must-be-pure.md [hunk 8]
```diff
@@ -307,35 +306,35 @@ function useIconStyle(icon) {
...
 ---
 
-## Values are immutable after being passed to JSX {/*values-are-immutable-after-being-passed-to-jsx*/}
+## Nilai tidak dapat diubah setelah diteruskan ke JSX {/*values-are-immutable-after-being-passed-to-jsx*/}
 
-Don't mutate values after they've been used in JSX. Move the mutation before the JSX is created.
+Jangan melakukan mutasi nilai setelah nilai tersebut digunakan dalam JSX. Pindahkan mutasi sebelum JSX dibuat.
 
-When you use JSX in an expression, React may eagerly evaluate the JSX before the component finishes rendering. This means that mutating values after they've been passed to JSX can lead to outdated UIs, as React won't know to update the component's output.
+Ketika Anda menggunakan JSX dalam sebuah ekspresi, React mungkin akan mengevaluasi JSX sebelum komponen selesai di-*render*. Ini berarti bahwa mengubah nilai setelah nilai tersebut dioper ke JSX dapat menyebabkan UI yang sudah usang, Karena React tidak akan tahu untuk memperbarui keluaran komponen.
```
Comment (resir014) [inline src/content/reference/rules/components-and-hooks-must-be-pure.md:331]:
```suggestion
Ketika Anda menggunakan JSX dalam sebuah ekspresi, React mungkin akan mengevaluasi JSX sebelum komponen selesai di-*render*. Ini berarti bahwa mengubah nilai setelah nilai tersebut dioper ke JSX dapat menyebabkan UI yang sudah usang, karena React tidak akan tahu untuk memperbarui keluaran komponen.
```

capitalisation

**Pull request merged**